deltacat/compute/compactor_v2/utils/compaction_utils.py

```diff
@@ -0,0 +1,800 @@
+import numpy as np
+import functools
+import logging
+import ray
+import time
+import json
+
+import deltacat
+
+from deltacat.compute.compactor import (
+    HighWatermark,
+    PyArrowWriteResult,
+    RoundCompletionInfo,
+)
+from deltacat.aws import s3u as s3_utils
+from deltacat.compute.compactor.model.compactor_version import CompactorVersion
+from deltacat.compute.compactor_v2.model.evaluate_compaction_result import (
+    ExecutionCompactionResult,
+)
+from deltacat.compute.compactor_v2.model.merge_file_group import (
+    RemoteMergeFileGroupsProvider,
+)
+from deltacat.compute.compactor_v2.model.hash_bucket_input import HashBucketInput
+
+from deltacat import logs
+from deltacat.compute.compactor_v2.model.merge_input import MergeInput
+from deltacat.compute.compactor_v2.model.merge_result import MergeResult
+from deltacat.compute.compactor_v2.model.hash_bucket_result import HashBucketResult
+from deltacat.compute.compactor.model.materialize_result import MaterializeResult
+from deltacat.compute.compactor_v2.utils.merge import (
+    generate_local_merge_input,
+)
+from deltacat.compute.compactor_v2.utils.task_options import (
+    hash_bucket_resource_options_provider,
+)
+from deltacat.compute.compactor.utils import round_completion_file as rcf
+from deltacat.compute.compactor import DeltaAnnotated
+from deltacat.compute.compactor_v2.utils.delta import contains_delete_deltas
+from deltacat.compute.compactor_v2.deletes.delete_strategy import (
+    DeleteStrategy,
+)
+from deltacat.compute.compactor_v2.deletes.delete_file_envelope import (
+    DeleteFileEnvelope,
+)
+from deltacat.utils.resources import (
+    get_current_process_peak_memory_usage_in_bytes,
+)
+from deltacat.compute.compactor_v2.deletes.utils import prepare_deletes
+
+from deltacat.storage import (
+    Delta,
+    DeltaType,
+    DeltaLocator,
+    Stream,
+    StreamLocator,
+    Partition,
+    Manifest,
+)
+from deltacat.compute.compactor.model.compact_partition_params import (
+    CompactPartitionParams,
+)
+from deltacat.utils.ray_utils.concurrency import (
+    invoke_parallel,
+    task_resource_options_provider,
+)
+from deltacat.compute.compactor_v2.steps import merge as mg
+from deltacat.compute.compactor_v2.steps import hash_bucket as hb
+from deltacat.compute.compactor_v2.utils import io
+
+from typing import List, Optional
+from collections import defaultdict
+from deltacat.compute.compactor.model.compaction_session_audit_info import (
+    CompactionSessionAuditInfo,
+)
+from deltacat.compute.compactor_v2.utils.task_options import (
+    merge_resource_options_provider,
+    local_merge_resource_options_provider,
+)
+
+
+logger = logs.configure_deltacat_logger(logging.getLogger(__name__))
+
+
+def _start_execute_compaction(
```

Comment:
Method name could perhaps be more descriptive --
We're not executing the compaction here, so maybe `_fetch_compaction_metadata` or `_prepare_for_compaction`.

We could also use either a docstring for this method or more explanation at this method's call site in compaction_session.

We should also add type annotations for the returned tuple.
You could consider defining a dataclass for the returned tuple, but it might be unnecessary.